<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using plotRegulome • plotRegulome</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using plotRegulome">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotRegulome</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/using_plotRegulome.html">Using plotRegulome</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="using_plotRegulome_files/kePrint-0.0.1/kePrint.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Using plotRegulome</h1>
                        <h4 class="author">Mireia Ramos-Rodríguez</h4>
            
      
      
      <div class="hidden name"><code>using_plotRegulome.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <code>plotRegulome</code> package has been developed with the idea of helping users produce and modify the <a href="http://isletregulome.com/isletregulome/">Islet Regulome Browser</a> plots.</p>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(plotRegulome)

## Download IRB database
<span class="co"># dir &lt;- "~/data/"</span>
<span class="co"># path &lt;- downloadIRB(output_dir=dir) # Download IRB data, returns path for database</span>

## Define parameters for plotting
coordinates &lt;-<span class="st"> "chr5:50678921-51318155"</span>
snps_dataset &lt;-<span class="st"> "diagram"</span>
contacts_dataset &lt;-<span class="st"> "ISL1"</span>
maps_dataset &lt;-<span class="st"> "openChromatinClasses"</span>
clusters_dataset &lt;-<span class="st"> "enhancerClusters"</span>
tfs_dataset &lt;-<span class="st"> "adult"</span>

## Call ploRegulome function
<span class="kw"><a href="../reference/plotRegulome.html">plotRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
             <span class="dt">snps_dataset=</span>snps_dataset,
             <span class="dt">contacts_dataset=</span>contacts_dataset,
             <span class="dt">maps_dataset=</span>maps_dataset,
             <span class="dt">clusters_dataset=</span>clusters_dataset,
             <span class="dt">tfs_dataset=</span>tfs_dataset,
             <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-2-1.png" width="1056"></p>
</div>
<div id="datasets" class="section level1">
<h1 class="hasAnchor">
<a href="#datasets" class="anchor"></a>Datasets</h1>
<p>In order to download all necessary datasets to plot islet data with this package, there is a wrapper function that does it for you. You just need to provide the path where the datasets for the IRB need to be stored.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">dir &lt;-<span class="st"> "~/data/"</span>

<span class="kw"><a href="../reference/downloadIRB.html">downloadIRB</a></span>(<span class="dt">output_dir=</span>dir) <span class="co"># Download IRB data</span></code></pre></div>
<p>You will see that the donwloaded data has the following structure:</p>
<pre><code>IRB_database
|-- code
|-- hg19
|   |-- clusters
|   |-- genes
|   |-- maps
|   |-- snps
|   |-- tfs
|   `-- virtual4c
|       |-- 1
|       |-- 2
|       |-- 3
|       |-- 4
|       |-- 5
|       |-- 6
|       |-- 7
|       |-- 8
|       `-- 9
`-- shared
</code></pre>
<p>Each folder inside <code>hg19</code> contains the datasets related to selectable data, summarized in the table below:</p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<caption style="font-size: initial !important;">
Selectable datasets for plotting.
</caption>
<thead><tr>
<th style="text-align:left;">
Name
</th>
<th style="text-align:left;">
Dataset
</th>
<th style="text-align:left;">
Reference
</th>
</tr></thead>
<tbody>
<tr grouplength="5">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Clusters</strong>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Enhancer Hubs
</td>
<td style="text-align:left;">
enhancerHubs
</td>
<td style="text-align:left;">
<span class="citation">Miguel-Escalada et al. (2019)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Super Enhancers
</td>
<td style="text-align:left;">
superEnhancers
</td>
<td style="text-align:left;">
<span class="citation">Miguel-Escalada et al. (2019)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Enhancer Clusters
</td>
<td style="text-align:left;">
enhancerClusters
</td>
<td style="text-align:left;">
<span class="citation">Pasquali et al. (2014)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Stretch Enhancers
</td>
<td style="text-align:left;">
stretchEnhancers
</td>
<td style="text-align:left;">
<span class="citation">S. C. J. Parker et al. (2013)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
COREs
</td>
<td style="text-align:left;">
cores
</td>
<td style="text-align:left;">
<span class="citation">K. J. Gaulton et al. (2010)</span>
</td>
</tr>
<tr grouplength="4">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>Maps</strong>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Adult Islets - Chromatin Classes
</td>
<td style="text-align:left;">
chromatinClassesReduced
</td>
<td style="text-align:left;">
<span class="citation">Miguel-Escalada et al. (2019)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Adult Islets - Chromatin Classes
</td>
<td style="text-align:left;">
openChromatinClasses
</td>
<td style="text-align:left;">
<span class="citation">Pasquali et al. (2014)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Pancreatic Progenitors
</td>
<td style="text-align:left;">
progenitors
</td>
<td style="text-align:left;">
<span class="citation">Cebola et al. (2015)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Adult Islets - Chromatin States
</td>
<td style="text-align:left;">
chromatinStates
</td>
<td style="text-align:left;">
<span class="citation">S. C. J. Parker et al. (2013)</span>
</td>
</tr>
<tr grouplength="3">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>SNPs</strong>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
70KforT2D
</td>
<td style="text-align:left;">
70KforT2D
</td>
<td style="text-align:left;">
<span class="citation">S. Bonàs-Guarch et al. (2018)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Diagram
</td>
<td style="text-align:left;">
diagram
</td>
<td style="text-align:left;">
<span class="citation">A. P. Morris et al. (2012)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Magic
</td>
<td style="text-align:left;">
magic
</td>
<td style="text-align:left;">
<span class="citation">R. A. Scott et al. (2012)</span>
</td>
</tr>
<tr grouplength="3">
<td colspan="3" style="border-bottom: 1px solid;">
<strong>TFs</strong>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Adult Islets <e2> Structural </e2>
</td>
<td style="text-align:left;">
structure
</td>
<td style="text-align:left;">
<span class="citation">Miguel-Escalada et al. (2019)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Pancreatic Progenitors
</td>
<td style="text-align:left;">
progenitors
</td>
<td style="text-align:left;">
<span class="citation">Cebola et al. (2015)</span>
</td>
</tr>
<tr>
<td style="text-align:left; padding-left: 2em;" indentlevel="1">
Adult Islets <e2> Tissue-specific </e2>
</td>
<td style="text-align:left;">
adult
</td>
<td style="text-align:left;">
<span class="citation">Pasquali et al. (2014)</span>
</td>
</tr>
</tbody>
</table>
<p>The name used for appropriatly calling datasets are listed in the column <code>Dataset</code> and in the documentation of <code>plotRegulome</code>.</p>
<p>Additionally, this package also includes gene annotation datasets (<code>genes</code> folder) that are shown by default:</p>
<table class="table table table-striped table-hover table-condensed table-responsive">
<caption style="font-size: initial !important;">
Gene annotation datasets
</caption>
<thead><tr>
<th style="text-align:left;">
Name
</th>
<th style="text-align:left;">
Dataset
</th>
<th style="text-align:left;">
Reference
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Gene annotation
</td>
<td style="text-align:left;">
gene_annotation_ensemblv75
</td>
<td style="text-align:left;">
<span class="citation">Zerbino et al. (2018)</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
RNA-seq
</td>
<td style="text-align:left;">
gene_moran
</td>
<td style="text-align:left;">
<span class="citation">Morán et al. (2012)</span>
</td>
</tr>
<tr>
<td style="text-align:left;">
lncRNA
</td>
<td style="text-align:left;">
lncRNA
</td>
<td style="text-align:left;">
<span class="citation">I. Akerman et al. (2017)</span>
</td>
</tr>
</tbody>
</table>
<p>Finally, virtual 4C data from <span class="citation">Miguel-Escalada et al. (2019)</span> can be found in <code>virtual4c</code> folder. Each bait is named with a numerical identifier and they are divided in folders according to the first number of the ID. This is done to prevent having t0o many files in one directory. The correspondence of bait IDs and bait names can be found in <code>virtual4c/baitID_keyTable.rda</code>.</p>
</div>
<div id="plot-description" class="section level1">
<h1 class="hasAnchor">
<a href="#plot-description" class="anchor"></a>Plot description</h1>
<div class="figure">
<img src="http://isletregulome.com/isletregulome/example_IRBPlot.png">
</div>
<p>The plot illustrates regulatory regions, transcription factors binding sites, chromatin interaction data and GWAS variants in which the sequence of the base genome is represented on the horizontal axis.</p>
<p>In the upper part of the plot a green line on the chromosome ideogram reflects the portion of the chromosome displayed. Each dot represents a genomic variant, being the color intensity of the dot proportional to -Log10 p-value of association, as indicated on the side of the plot. The rs# ID depicts the top associated variants in the locus.</p>
<p>The virtual 4C data is visualized as an interaction frequency histogram plot, where the bait/viewpoint is depicted as a black triangle region. A distal peak in the signal indicates that there is a chromatin interaction event with the bait/viewpoint. The interaction significance is depicted by a color code reflecting the <a href="http://regulatorygenomicsgroup.org/chicago">CHiCAGO scores</a>. Such chromatin interaction maps were extracted from promoter capture HiC data centered on the queried bait region, hence, virtual 4C.</p>
<p>The black box in the middle of the plot contains vertical colored bands depicting different chromatin states, open chromatin classes or regulatory elements as described in the legend above the plot. Black lines connecting the circles (each representing a different transcription factor) to the black box, point to the genomic location of each transcription factor binding site. The color intensity of such lines is proportional to the number of co-bound transcription factors.</p>
<p>Annotated genes are depicted as horizontal gray lines at the bottom of the plot. Boxes along the line correspond to positions of coding exons. The color of the box indicates the type of annotation: black boxes indicate lncRNAs, grey boxes indicate genes and purple boxes represent islet-specific genes.</p>
</div>
<div id="plotting-the-islet-regulome-plot" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-the-islet-regulome-plot" class="anchor"></a>Plotting the Islet Regulome plot</h1>
<p>For a detailed description of all plot elements you can visit <a href="isletregulome.com" class="uri">isletregulome.com</a> and go to Info &gt; Description of the plot or read the section <em>Plot description</em> below.</p>
<p>In order to plot an Islet Regulome image, you will need to provide the coordinates of the region you want to plot, together with the names of the datasets you want to show. Please check <code><a href="../reference/plotRegulome.html">?plotRegulome</a></code> or the <strong>Datasets</strong> section to see all available datasets and their code names.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(plotRegulome)

coordinates &lt;-<span class="st"> "chr13:28354157-28550368"</span>
snps_dataset &lt;-<span class="st"> "70KforT2D"</span>
contacts_dataset &lt;-<span class="st"> "PDX1"</span>
maps_dataset &lt;-<span class="st"> "chromatinClassesReduced"</span>
clusters_dataset &lt;-<span class="st"> "enhancerHubs"</span>
tfs_dataset &lt;-<span class="st"> "structure"</span>

<span class="kw"><a href="../reference/plotRegulome.html">plotRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
             <span class="dt">snps_dataset=</span>snps_dataset,
             <span class="dt">contacts_dataset=</span>contacts_dataset,
             <span class="dt">maps_dataset=</span>maps_dataset,
             <span class="dt">clusters_dataset=</span>clusters_dataset,
             <span class="dt">tfs_dataset=</span>tfs_dataset,
             <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-5-1.png" width="1056"></p>
<p>You can save the Islet Regulome plots using the <code>ggsave</code> function from the <code>ggplot2</code> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">ggsave</span>(<span class="dt">file=</span><span class="st">"example_plotRegulome.png"</span>,
       <span class="dt">width=</span><span class="dv">11</span>, <span class="dt">height=</span><span class="dv">6</span>)</code></pre></div>
<div id="customizing-islet-regulome-plot-colors" class="section level2">
<h2 class="hasAnchor">
<a href="#customizing-islet-regulome-plot-colors" class="anchor"></a>Customizing Islet Regulome plot colors</h2>
<p>Additional arguments to change the plotting color for the different elements can be used:</p>
<ul>
<li>
<code>snps_col</code>. Character with the name or the hex value for the darkest color to use when plotting SNPs. Default: <code>dark red</code>.</li>
<li>
<code>contacts_col</code>. Named list with the colors for each grouping of CHiCAGO scores. Default: <code><a href="https://rdrr.io/r/base/c.html">c("0"="grey", "3"="blue", "5"="dark orange")</a></code>
</li>
<li>
<code>cluster_col</code>. Character with the name or the hex value for color to use when plotting enhancer clusters. Default: <code>dark green</code>.</li>
<li>
<code>tfs_col</code>. Character with the name or the hex value for the TFs border. Default: “dark blue”.</li>
<li>
<code>genes_col</code>. Named list with the colors for each type of gene annotation, where <code>gene</code> refers to gene annotation, <code>lnc</code> to lncRNAs and <code>spec</code> to islet-specific genes. Default: <code><a href="https://rdrr.io/r/base/c.html">c(gene="dark grey", spec="darkorchid3", lnc="black")</a></code>
</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/plotRegulome.html">plotRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
             <span class="dt">snps_dataset=</span>snps_dataset,
             <span class="dt">snps_col=</span><span class="st">"mediumorchid4"</span>,
             <span class="dt">contacts_dataset=</span>contacts_dataset,
             <span class="dt">contacts_col=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"0"</span>=<span class="st">"dark grey"</span>, <span class="st">"3"</span>=<span class="st">"skyblue3"</span>, <span class="st">"5"</span>=<span class="st">"olivedrab4"</span>),
             <span class="dt">maps_dataset=</span>maps_dataset,
             <span class="dt">clusters_dataset=</span>clusters_dataset,
             <span class="dt">cluster_col=</span><span class="st">"dodgerblue4"</span>,
             <span class="dt">tfs_dataset=</span>tfs_dataset,
             <span class="dt">tfs_col=</span><span class="st">"violetred4"</span>,
             <span class="dt">genes_col=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"gene"</span>=<span class="st">"thistle3"</span>, <span class="st">"lnc"</span>=<span class="st">"orangered3"</span>, <span class="st">"spec"</span>=<span class="st">"yellow3"</span>),
             <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-7-1.png" width="1056"></p>
</div>
<div id="random-islet-regulome-plot-generator" class="section level2">
<h2 class="hasAnchor">
<a href="#random-islet-regulome-plot-generator" class="anchor"></a>Random Islet Regulome plot generator</h2>
<p>Finally, there is an additional argument to the <code>plotRegulome</code> function that generates random plots with random available datasets. For simplicity reasons, <code>contacts_dataset</code> are not used, as it is not likely that a bait in the selected region would be randomly selected.</p>
<p>Any specifications on datasets will be overwritten when <code>randomIRB=TRUE</code>. However, you can still specify colors for the different data as explained above.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">2384982</span>)
  
<span class="kw"><a href="../reference/plotRegulome.html">plotRegulome</a></span>(<span class="dt">randomIRB=</span>T,
             <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)
<span class="co">#&gt; ------------------------------------------------------</span>
<span class="co">#&gt; -- Random set generated at chr1:113559015-113630541 --</span>
<span class="co">#&gt; ------------------------------------------------------</span>
<span class="co">#&gt; Clusters: enhancerHubs</span>
<span class="co">#&gt; Maps: chromatinStates</span>
<span class="co">#&gt; SNPs: diagram</span>
<span class="co">#&gt; TFs: progenitors</span>
<span class="co">#&gt; ------------------------------------------------------</span></code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-8-1.png" width="1056"></p>
</div>
</div>
<div id="plotting-datasets-separately" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting-datasets-separately" class="anchor"></a>Plotting datasets separately</h1>
<p>In case you are interested in plotting one or a few datasets from the Islet Regulome, this package also contains the necessary functions to do so.</p>
<p>The first step is to create an object with tha dataset and type of data you want to plot. For this reason, there are a set for creator functions that will compute all necessary data to plot a specific dataset:</p>
<ul>
<li>
<code>create_snpsRegulome</code>. Generates a <code>snpsRegulome</code> object containing SNPs data in the region of interest.</li>
<li>
<code>create_contactsRegulome</code>. Produces a <code>contactsRegulome</code> object containing regions in contact with the bait of interest and their CHiCAGO scores.</li>
<li>
<code>create_mapsRegulome</code>. Produces a <code>mapsRegulome</code> object with the chromatin maps in the region of interest.</li>
<li>
<code>create_clustersRegulome</code>. Produces a <code>clustersRegulome</code> object with information of the enhancer clusters in the reigon of interest.</li>
<li>
<code>create_tfsRegulome</code>. Produces a <code>tfsRegulome</code> object with information on the TFBSin the region of interest.</li>
<li>
<code>create_genesRegulome</code>. Generates gene annotation in a <code>genesRegulome</code> object containing information of genes in the region, islet lncRNAs and their islet specificity.</li>
</ul>
<p>All of these functions take a <code>coordinates</code> argument, which is a <code>GRanges</code> object with the coordinates of the region you want to plot, and an <code>xxx_dataset</code> (except <code>create_genesRegulome</code>), which contains the dataset name as defined in the section <em>Datasets</em> above. Each of this function has additional arguments which you can check in their corresponding check pages (i.e <code><a href="../reference/create_snpsRegulome.html">?create_snpsRegulome</a></code>).</p>
<p>For plotting each dataset, you just need to use the <code><a href="https://rdrr.io/r/graphics/plot.html">plot()</a></code> function with the <code>xxxRegulome</code> object you just created. This will produce a list of ggplot layers that need to be added to a <code><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot()</a></code> function to generate the plot.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(ggplot2)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(plotRegulome)

coordinates &lt;-<span class="st"> "chr20:22461643-22666093"</span>

snps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_snpsRegulome.html">create_snpsRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
                            <span class="dt">snps_dataset=</span><span class="st">"magic"</span>,
                            <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)


<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>() +<span class="st"> </span><span class="kw"><a href="../reference/plotR.html">plotR</a></span>(snps)</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">contacts &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_contactsRegulome.html">create_contactsRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
                                    <span class="dt">contacts_dataset=</span><span class="st">"FOXA2"</span>,
                                    <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)

<span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>() +<span class="st"> </span><span class="kw"><a href="../reference/plotR.html">plotR</a></span>(contacts)</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Other types of dataset are of no much use when plot alone, such as <code>tfsRegulome</code> or <code>clustersRegulome</code>. You can also plot them together as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">enh &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_clustersRegulome.html">create_clustersRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
                               <span class="dt">clusters_dataset=</span><span class="st">"superEnhancers"</span>,
                               <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)

maps &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_mapsRegulome.html">create_mapsRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
                            <span class="dt">maps_dataset=</span><span class="st">"chromatinClassesReduced"</span>,
                            <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)

tfs &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_tfsRegulome.html">create_tfsRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
                          <span class="dt">tfs_dataset=</span><span class="st">"adult"</span>,
                          <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)


comb_plot &lt;-<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>() +<span class="st"> </span><span class="kw"><a href="../reference/plotR.html">plotR</a></span>(enh) +<span class="st"> </span><span class="kw"><a href="../reference/plotR.html">plotR</a></span>(maps) +<span class="st"> </span><span class="kw"><a href="../reference/plotR.html">plotR</a></span>(tfs) +<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/lims.html">ylim</a></span>(<span class="fl">0.4</span>,<span class="ot">NA</span>)
comb_plot</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>If you want to include the gene annotation, you should generate the plot separately and you can then combine it with the maps using <code>plot_grid</code> function from the <a href="https://cran.r-project.org/web/packages/cowplot/index.html"><code>cowplot</code></a> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/create_genesRegulome.html">create_genesRegulome</a></span>(<span class="dt">coordinates=</span>coordinates,
                              <span class="dt">path=</span><span class="st">"~/data/IRB_database/"</span>)

anno_plot &lt;-<span class="st"> </span><span class="kw"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span>() +<span class="st"> </span><span class="kw"><a href="../reference/plotR.html">plotR</a></span>(genes)

<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(cowplot)
cowplot::<span class="kw"><a href="https://rdrr.io/pkg/cowplot/man/plot_grid.html">plot_grid</a></span>(comb_plot +<span class="st"> </span><span class="kw"><a href="../reference/themeXblank.html">themeXblank</a></span>(),
                   anno_plot,
                   <span class="dt">ncol=</span><span class="dv">1</span>,
                   <span class="dt">align=</span><span class="st">"v"</span>,
                   <span class="dt">rel_heights =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0.6</span>, <span class="fl">0.4</span>))</code></pre></div>
<p><img src="using_plotRegulome_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Akerman2017">
<p>Akerman, Ildem, Zhidong Tu, Anthony Beucher, Delphine M.Y. Rolando, Claire Sauty-Colace, Marion Benazra, Nikolina Nakic, et al. 2017. “Human Pancreatic <span class="math inline">\(\beta\)</span> Cell lncRNAs Control Cell-Specific Regulatory Networks.” <em>Cell Metabolism</em> 25 (2): 400–411. doi:<a href="https://doi.org/10.1016/j.cmet.2016.11.016">10.1016/j.cmet.2016.11.016</a>.</p>
</div>
<div id="ref-Bonas-Guarch2018">
<p>Bonàs-Guarch, Sílvia, Marta Guindo-Martínez, Irene Miguel-Escalada, Niels Grarup, David Sebastian, Elias Rodriguez-Fos, Friman Sánchez, et al. 2018. “Re-analysis of public genetic data reveals a rare X-chromosomal variant associated with type 2 diabetes.” <em>Nature Communications</em> 9 (1). Nature Publishing Group: 321. doi:<a href="https://doi.org/10.1038/s41467-017-02380-9">10.1038/s41467-017-02380-9</a>.</p>
</div>
<div id="ref-Cebola2015">
<p>Cebola, Inês, Santiago A. Rodríguez-Seguí, Candy H-H. Cho, José Bessa, Meritxell Rovira, Mario Luengo, Mariya Chhatriwala, et al. 2015. “TEAD and YAP regulate the enhancer network of human embryonic pancreatic progenitors.” <em>Nature Cell Biology</em> 17 (5): 615–26. doi:<a href="https://doi.org/10.1038/ncb3160">10.1038/ncb3160</a>.</p>
</div>
<div id="ref-Gaulton2010">
<p>Gaulton, Kyle J, Takao Nammo, Lorenzo Pasquali, Jeremy M Simon, Paul G Giresi, Marie P Fogarty, Tami M Panhuis, et al. 2010. “A map of open chromatin in human pancreatic islets.” <em>Nature Genetics</em> 42 (3). Nature Publishing Group: 255–59. doi:<a href="https://doi.org/10.1038/ng.530">10.1038/ng.530</a>.</p>
</div>
<div id="ref-Miguel-Escalada2019">
<p>Miguel-Escalada, Irene, Silvia Bonàs-Guarch, Inês Cebola, Joan Ponsa-Cobas, Julen Mendieta-Esteban, Goutham Atla, Biola M. Javierre, et al. 2019. “Human pancreatic islet three-dimensional chromatin architecture provides insights into the genetics of type 2 diabetes.” <em>Nature Genetics</em> 51 (7). Nature Publishing Group: 1137–48. doi:<a href="https://doi.org/10.1038/s41588-019-0457-0">10.1038/s41588-019-0457-0</a>.</p>
</div>
<div id="ref-Moran2012">
<p>Morán, Ignasi, İldem Akerman, Martijn van de Bunt, Ruiyu Xie, Marion Benazra, Takao Nammo, Luis Arnes, et al. 2012. “Human <span class="math inline">\(\beta\)</span> Cell Transcriptome Analysis Uncovers lncRNAs That Are Tissue-Specific, Dynamically Regulated, and Abnormally Expressed in Type 2 Diabetes.” <em>Cell Metabolism</em> 16 (4): 435–48. doi:<a href="https://doi.org/10.1016/j.cmet.2012.08.010">10.1016/j.cmet.2012.08.010</a>.</p>
</div>
<div id="ref-Morris2012">
<p>Morris, Andrew P, Benjamin F Voight, Tanya M Teslovich, Teresa Ferreira, Ayellet V Segrè, Valgerdur Steinthorsdottir, Rona J Strawbridge, et al. 2012. “Large-scale association analysis provides insights into the genetic architecture and pathophysiology of type 2 diabetes.” <em>Nature Genetics</em> 44 (9). Nature Publishing Group: 981–90. doi:<a href="https://doi.org/10.1038/ng.2383">10.1038/ng.2383</a>.</p>
</div>
<div id="ref-Parker2013">
<p>Parker, Stephen C J, Michael L Stitzel, D Leland Taylor, Jose Miguel Orozco, Michael R Erdos, Jennifer A Akiyama, Kelly Lammerts van Bueren, et al. 2013. “Chromatin stretch enhancer states drive cell-specific gene regulation and harbor human disease risk variants.” <em>Proceedings of the National Academy of Sciences</em> 110 (44): 17921–6. doi:<a href="https://doi.org/10.1073/pnas.1317023110">10.1073/pnas.1317023110</a>.</p>
</div>
<div id="ref-Pasquali2014">
<p>Pasquali, Lorenzo, Kyle J. Gaulton, Santiago A. Rodríguez-Seguí, Loris Mularoni, Irene Miguel-Escalada, İldem Ildem Akerman, Juan J. Tena, et al. 2014. “Pancreatic islet enhancer clusters enriched in type 2 diabetes risk-associated variants.” <em>Nature Genetics</em> 46 (2): 136–43. doi:<a href="https://doi.org/10.1038/ng.2870">10.1038/ng.2870</a>.</p>
</div>
<div id="ref-Scott2012">
<p>Scott, Robert A, Vasiliki Lagou, Ryan P Welch, Eleanor Wheeler, May E Montasser, Jian’an Luan, Reedik Mägi, et al. 2012. “Large-scale association analyses identify new loci influencing glycemic traits and provide insight into the underlying biological pathways.” <em>Nature Genetics</em> 44 (9): 991–1005. doi:<a href="https://doi.org/10.1038/ng.2385">10.1038/ng.2385</a>.</p>
</div>
<div id="ref-Zerbino2018">
<p>Zerbino, Daniel R, Premanand Achuthan, Wasiu Akanni, M Ridwan Amode, Daniel Barrell, Jyothish Bhai, Konstantinos Billis, et al. 2018. “Ensembl 2018.” <em>Nucleic Acids Research</em> 46 (D1). Narnia: D754–D761. doi:<a href="https://doi.org/10.1093/nar/gkx1098">10.1093/nar/gkx1098</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#quick-start">Quick start</a></li>
      <li><a href="#datasets">Datasets</a></li>
      <li><a href="#plot-description">Plot description</a></li>
      <li>
<a href="#plotting-the-islet-regulome-plot">Plotting the Islet Regulome plot</a><ul class="nav nav-pills nav-stacked">
<li><a href="#customizing-islet-regulome-plot-colors">Customizing Islet Regulome plot colors</a></li>
      <li><a href="#random-islet-regulome-plot-generator">Random Islet Regulome plot generator</a></li>
      </ul>
</li>
      <li><a href="#plotting-datasets-separately">Plotting datasets separately</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
