<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title><code>plotRegulome</code> package — plotRegulome • plotRegulome</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/yeti/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="<code>plotRegulome</code> package — plotRegulome" />

<meta property="og:description" content="Plotting Islet Regulome data
By providing the coordinates and IRB datasets of interest, it will produce the characteristic
 Islet Regulome Broser plot." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plotRegulome</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/using_plotRegulome.html">Using plotRegulome</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1><code>plotRegulome</code> package</h1>
    
    <div class="hidden name"><code>plotRegulome.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Plotting Islet Regulome data</p>
<p>By providing the coordinates and IRB datasets of interest, it will produce the characteristic
 Islet Regulome Broser plot.</p>
    
    </div>

    <pre class="usage"><span class='fu'>plotRegulome</span>(<span class='no'>coordinates</span>, <span class='kw'>snps_dataset</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>snps_col</span> <span class='kw'>=</span> <span class='st'>"dark red"</span>,
  <span class='kw'>contacts_dataset</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>contacts_col</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>`0`</span> <span class='kw'>=</span> <span class='st'>"grey"</span>, <span class='kw'>`3`</span> <span class='kw'>=</span> <span class='st'>"blue"</span>,
  <span class='kw'>`5`</span> <span class='kw'>=</span> <span class='st'>"dark orange"</span>), <span class='kw'>maps_dataset</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>clusters_dataset</span> <span class='kw'>=</span> <span class='st'>""</span>,
  <span class='kw'>cluster_col</span> <span class='kw'>=</span> <span class='st'>"dark green"</span>, <span class='kw'>tfs_dataset</span> <span class='kw'>=</span> <span class='st'>""</span>,
  <span class='kw'>tfs_col</span> <span class='kw'>=</span> <span class='st'>"dark blue"</span>, <span class='kw'>genes_col</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>gene</span> <span class='kw'>=</span> <span class='st'>"dark grey"</span>, <span class='kw'>spec</span> <span class='kw'>=</span>
  <span class='st'>"darkorchid3"</span>, <span class='kw'>lnc</span> <span class='kw'>=</span> <span class='st'>"black"</span>), <span class='kw'>showLongestTranscript</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>randomIRB</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>genome</span> <span class='kw'>=</span> <span class='st'>"hg19"</span>, <span class='kw'>path</span> <span class='kw'>=</span> <span class='st'>"~/data/IRB/"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>coordinates</th>
      <td><p>Either a <code>GRanges</code> object, data.frame or character string
(i.e. <code>chr11:17227951-17589050</code>) indicating the coordinates for the region the user wants
to plot.</p></td>
    </tr>
    <tr>
      <th>snps_dataset</th>
      <td><p>GWAS SNPs dataset name to use for the analysis. The value should be one
of the defined below (column <em>Dataset</em>). Defaults to <code>""</code>, which will produce an
empty plot.</p><table class='table'>
<tr><td><strong>Name</strong></td><td><strong>Dataset</strong></td><td><strong>Reference</strong></td></tr>
<tr><td>70KforT2D</td><td>70KforT2D</td><td>S. Bonàs-Guarch et al. (2018)</td></tr>
<tr><td>Diagram</td><td>diagram</td><td>A. P. Morris et al. (2012)</td></tr>
<tr><td>Magic</td><td>magic</td><td>R. A. Scott et al. (2012)</td></tr>
</table>
</td>
    </tr>
    <tr>
      <th>snps_col</th>
      <td><p>Color for the GWAS SNPs. Default: "dark red"</p></td>
    </tr>
    <tr>
      <th>contacts_dataset</th>
      <td><p>BaitID or bait gene name for the virtual 4C data (Miguel-Escalada
et al. (2019)) to plot.</p></td>
    </tr>
    <tr>
      <th>contacts_col</th>
      <td><p>Named vector ("0", "3" and "5"), with the colors for each group of CHiCAGO
contacts.</p></td>
    </tr>
    <tr>
      <th>maps_dataset</th>
      <td><p>Name of the chromatin maps to plot. The value should be one
of the defined below (column <em>Dataset</em>). Defaults to <code>""</code>, which will produce an
empty plot.</p><table class='table'>
<tr><td><strong>Name</strong></td><td><strong>Dataset</strong></td><td><strong>Reference</strong></td></tr>
<tr><td>Adult Islets - Chromatin Classes</td><td>chromatinClassesReduced</td><td>Miguel-Escalada et al. (2019)</td></tr>
<tr><td>Adult Islets -  Chromatin Classes</td><td>openChromatinClasses</td><td>Pasquali et al. (2014)</td></tr>
<tr><td>Pancreatic Progenitors</td><td>progenitors</td><td>Cebola et al. (2015)</td></tr>
<tr><td>Adult Islets - Chromatin States</td><td>chromatinStates</td><td>Parker et al. (2013)</td></tr>
</table>
</td>
    </tr>
    <tr>
      <th>clusters_dataset</th>
      <td><p>Name of the enhncer cluster data to plot. The value should be one
of the defined below (column <em>Dataset</em>). Defaults to <code>""</code>, which will produce an
empty plot.</p><table class='table'>
<tr><td><strong>Name</strong></td><td><strong>Dataset</strong></td><td><strong>Reference</strong></td></tr>
<tr><td>Enhancer Hubs</td><td>enhancerHubs</td><td>Miguel-Escalada et al. (2019)</td></tr>
<tr><td>Super Enhancers</td><td>superEnhancers</td><td>Miguel-Escalada et al. (2019)</td></tr>
<tr><td>Enhancer Clusters</td><td>enhancerClusters</td><td>Pasquali et al. (2014)</td></tr>
<tr><td>Stretch Enhancers</td><td>stretchEnhancers</td><td>Parker et al. (2013)</td></tr>
<tr><td>COREs</td><td>cores</td><td>K. J. Gaulton et al. (2010)</td></tr>
</table>
</td>
    </tr>
    <tr>
      <th>cluster_col</th>
      <td><p>Color for the clusters. Default: "dark green"</p></td>
    </tr>
    <tr>
      <th>tfs_dataset</th>
      <td><p>Name of the TF dataset to plot. The value should be one
of the defined below (column <em>Dataset</em>). Defaults to <code>""</code>, which will produce an
empty plot.</p><table class='table'>
<tr><td><strong>Name</strong></td><td><strong>Dataset</strong></td><td><strong>Reference</strong></td></tr>
<tr><td>Adult Islets – Structural</td><td>structure</td><td>Miguel-Escalada et al. (2019)</td></tr>
<tr><td>Pancreatic Progenitors</td><td>progenitors</td><td>Cebola et al. (2015)</td></tr>
<tr><td>Adult Islets – Tissue-specific</td><td>adult</td><td>Pasquali et al. (2014)</td></tr>
</table>
</td>
    </tr>
    <tr>
      <th>tfs_col</th>
      <td><p>Color for the TFs circle border.</p></td>
    </tr>
    <tr>
      <th>genes_col</th>
      <td><p>Named character vector ("gene", "lnc" and "spec") with the colors for each type of feture plotted in the gene
annotation track.</p></td>
    </tr>
    <tr>
      <th>showLongestTranscript</th>
      <td><p>When plotting gene data, set to TRUE (default) if you want to reduce the
number of transcripts by only plotting the longest transcript per gene. If set to FALSE, will plot all
available transcripts.</p></td>
    </tr>
    <tr>
      <th>randomIRB</th>
      <td><p>Generate random combinations of IRB datasets.</p></td>
    </tr>
    <tr>
      <th>genome</th>
      <td><p>Character string indicating the genome for the coordinates. Default: hg19.</p></td>
    </tr>
    <tr>
      <th>path</th>
      <td><p>Path containing the genomes folder (for example "hg19").
Default: "~/data/IRB/"</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A ggplot2 object that can be plotted or saved with ggsave. Recomended device size in
inches is: <code>width=11, height=6</code></p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>See the README on
<a href='https://github.com/mireia-bioinfo/plotRegulome#readme'>GitHub</a></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fl'>FALSE</span>) {
    <span class='no'>coordinates</span> <span class='kw'>&lt;-</span> <span class='st'>"chr5:131817301-131826490"</span>
    <span class='no'>snps_dataset</span> <span class='kw'>&lt;-</span> <span class='st'>"70KforT2D"</span>
    <span class='no'>contacts_dataset</span> <span class='kw'>&lt;-</span> <span class='fl'>570519</span> <span class='co'>#183446</span>
    <span class='no'>maps_dataset</span> <span class='kw'>&lt;-</span> <span class='st'>"progenitors"</span>
    <span class='no'>clusters_dataset</span> <span class='kw'>&lt;-</span> <span class='st'>""</span>
    <span class='no'>tfs_dataset</span> <span class='kw'>&lt;-</span> <span class='st'>""</span>
    <span class='no'>path</span> <span class='kw'>&lt;-</span> <span class='st'>"~/data/IRB/"</span>

    <span class='fu'>plotRegulome</span>(<span class='kw'>coordinates</span><span class='kw'>=</span><span class='no'>coordinates</span>,
                 <span class='kw'>snps_dataset</span><span class='kw'>=</span><span class='no'>snps_dataset</span>,
                 <span class='kw'>contacts_dataset</span><span class='kw'>=</span><span class='no'>contacts_dataset</span>,
                 <span class='kw'>maps_dataset</span><span class='kw'>=</span><span class='no'>maps_dataset</span>,
                 <span class='kw'>clusters_dataset</span><span class='kw'>=</span><span class='no'>clusters_dataset</span>,
                 <span class='kw'>tfs_dataset</span><span class='kw'>=</span><span class='no'>tfs_dataset</span>,
                 <span class='kw'>path</span><span class='kw'>=</span><span class='no'>path</span>)
}</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Mireia Ramos-Rodriguez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


